TcgPocket/Shared/PagedResult/QueryableExtensions.cs

```diff
@@ -167,6 +107,48 @@ public static PagedResult<TEntity> HandleNullPageDto<TEntity>(this IQueryable<TE
         return pagedEntity;
     }
 
+    public static IQueryable<TEntity> FilterBy<TEntity, TFilter>(this IQueryable<TEntity> query, TFilter filter)
+    where TEntity : class
+    where TFilter : class
+    {
+        // get object fields
+        var entityFields = typeof(TEntity).GetProperties();
+        var filterFields = typeof(TFilter).GetProperties();
+
+        // get common fields where filter value is not null 
+        var commonFields = filterFields
+            .Where(x => entityFields.Any(y => y.Name == x.Name))
+            .Where(fields => fields.GetValue(filter, null) is not null)
+            .ToList();
+
+        // setup predicate and type expression
+        var entityTypeExpression = Expression.Parameter(typeof(TEntity));
+        Expression? expression = null;
+
+        commonFields.ForEach(field =>
+        {
+            // get entity field and value
+            var entityField = Expression.PropertyOrField(entityTypeExpression, field.Name);
+            var entityFieldType = entityField.Type;
+
+            // convert raw field value to desired type (probably not necessary but it makes me feel safe)
+            var rawFiledValue = field.GetValue(filter);
+            var convertedFieldValue = Convert.ChangeType(rawFiledValue, entityFieldType);
+            
+            var filterExpression = 
+                Expression.Equal(entityField, Expression.Constant(convertedFieldValue, entityFieldType));
```

Comment:
```suggestion
            Expression filterExpression = null;

            if (field.PropertyType == typeof(string))
            {
                filterExpression =
                    Expression.Call(entityField, "Contains", null, Expression.Constant(convertedFieldValue));
            }
            else
            {
                filterExpression =
                    Expression.Equal(entityField, Expression.Constant(convertedFieldValue, entityFieldType));
            }
```
Would suggest adding something like this for fields that are string types so that we can filter by portions of string rather an exact matches.

Reply:
beautiful 😊